ratis-server-api/src/main/java/org/apache/ratis/server/raftlog/RaftLog.java

```diff
@@ -58,8 +59,21 @@ default boolean contains(TermIndex ti) {
   /**
    * @return null if the log entry is not found in this log;
    *         otherwise, return the log entry corresponding to the given index.
+   * @deprecated use {@link RaftLog#getWithRef(long)} instead.
    */
-  LogEntryProto get(long index) throws RaftLogIOException;
+  @Deprecated
+  default LogEntryProto get(long index) throws RaftLogIOException {
+    ReferenceCountedObject<LogEntryProto> ref = getWithRef(index);
+    return ref != null ? ref.get() : null;
+  }
+
+  /**
+   * @return a ReferenceCountedObject to the  log entry corresponding to the given log index if it exists.
+   * otherwise, null.
+   */
+  default ReferenceCountedObject<LogEntryProto> getWithRef(long index) throws RaftLogIOException {
+    return ReferenceCountedObject.wrap(get(index));
+  }
```

Comment:
Similar to `TransactionContext.getLogEntryRef()` below.  We should think about if the caller should invoke `retain()` before using it.  If `retain()` is not needed, when is it safe to call `get()`.

https://github.com/apache/ratis/blob/1bd451046b2d33fe063712d7e811262bef580a4e/ratis-server-api/src/main/java/org/apache/ratis/statemachine/TransactionContext.java#L115-L130

Reply:
`TransactionContext.getLogEntryRef()` will be protected by a retain/release pair from the framework. So, the StateMachine code doesn't need to call retain/release unless extending the LogEntry usage outside the method scope. It's where the javadoc above from.

```
// in StateMachineUpdater.
final ReferenceCountedObject<LogEntryProto> next = raftLog.getWithRef(nextIndex);
next.retain();
try {
  server.applyLogToStateMachine(next);
} finally {
  next.release();
}
```